cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskRetentionManager.java

```diff
@@ -0,0 +1,214 @@
+                                                                  : config.getInt(KafkaCruiseControlConfig.
+                                                                                      MAX_CACHED_COMPLETED_MONITOR_USER_TASKS_CONFIG)));
+    _endPointRetentionByType.put(ADMIN_KAFKA,
+                                 new RetentionInfo(ADMIN_KAFKA.endPoints(),
+                                                   config == null ? defaultUserTaskRetentionTimeMs
+                                                                  : config.getLong(KafkaCruiseControlConfig.
+                                                                                       COMPLETED_ADMIN_USER_TASK_RETENTION_TIME_MS_CONFIG),
+                                                   config == null ? defaultUserTaskRetentionLimit
+                                                                  : config.getInt(KafkaCruiseControlConfig.
+                                                                                      MAX_CACHED_COMPLETED_ADMIN_USER_TASKS_CONFIG)));
+    _endPointRetentionByType.put(ADMIN_CRUISE_CONTROL,
+                                 new RetentionInfo(ADMIN_CRUISE_CONTROL.endPoints(),
+                                                   config == null ? defaultUserTaskRetentionTimeMs
+                                                                  : config.getLong(KafkaCruiseControlConfig.
+                                                                                       COMPLETED_ADMIN_USER_TASK_RETENTION_TIME_MS_CONFIG),
+                                                   config == null ? defaultUserTaskRetentionLimit
+                                                                  : config.getInt(KafkaCruiseControlConfig.
+                                                                                      MAX_CACHED_COMPLETED_ADMIN_USER_TASKS_CONFIG)));
+    _time = time;
+  }
+
+  /**
+   * Add {@link UUID} associated with a UserTask to corresponding category in {@link UserTaskRetentionManager}
+   * @param uuid {@link UUID} to be added
+   * @param endPoint {@link EndPoint} used to determine category of UserTask
+   */
+  public void addUserTaskUUID(UUID uuid, EndPoint endPoint) {
+    RetentionInfo retentionInfo = retentionInfoFromEndPoint(endPoint);
+    if (retentionInfo != null) {
+      retentionInfo.addUUID(uuid);
+    }
+  }
+
+  private RetentionInfo retentionInfoFromEndPoint(EndPoint endPoint) {
+    for (RetentionInfo retentionInfo : _endPointRetentionByType.values()) {
+      if (retentionInfo.endPoints().contains(endPoint)) {
+        return retentionInfo;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Removes UserTasks for a certain {@link RetentionType} that has exceeded retention time for that type.
+   * @param idToUserTask Map of @{link UUID} to UserTasks
+   */
+  public void removeOldUserTasks(Map<UUID, UserTaskManager.UserTaskInfo> idToUserTask) {
```

Comment:
Instead of always pass completed task list to `removeOldUserTasks()` and  `removeExtraUserTasks()` , I feel a better way is to totally move completed task list into `UserTaskRetentionManager` and let `UserTaskRetentionManager` to maintain it. 
So `UserTaskManager` only maintains active task list, once a task is completed, just give the task to `UserTaskRetentionManager`, `UserTaskRetentionManager` will take care of everything about retention.

Reply:
I think it's better to keep list of completed tasks in UserTaskManager, along with ACTIVE/EXECUTING etc tasks because it's more consistent. I'm thinking UserTaskRetentionManager as a kind of helper that acts on a list.